### ISSUE TITLE
Allow artifacts to be non-tarball files?

### ISSUE THREAD

**Issue by thomasjm** [JuliaLang/julia#38599]
The `Pkg.Artifacts` system is awesome, but it seems to assume that every artifact is a zip/tar archive that needs to be decompressed after it's downloaded. (Discussed on Discourse [here](https://discourse.julialang.org/t/does-an-artifact-download-always-need-to-be-a-tarball-zipped-file/33964).)

I think it would be convenient to be able to download arbitrary kinds of files as artifacts. For example, I'm looking at converting [PlotlyJS.jl](https://github.com/JuliaPlots/PlotlyJS.jl) to use artifacts. It currently needs to download 2 things: a JSON file and a minified Javascript file. Both of these files can be readily downloaded from the `plot.ly` website or from a CDN, but it's hard to get them inside a tarball. (Technically I can find them both in a huge 93MB release tarball as part of `plotly.js`, but that requires downloading a lot of unnecessary stuff.)

This situation seems to require that package maintainers create and maintain their own source of tarballs, such as putting them in GitHub releases, when I'd rather just download the files from where they naturally come from.

**Comment by Keno** [comment]
The plan was to start serving artifacts from the PkgServers in case upstreams go away or are otherwise unavailable. As such, standardizing on the format of artifacts is important to make sure that they can be reprocessed by the package server for delivery.

**Comment by thomasjm** [comment]
Huh. What would "reprocessing" mean here? It sounds like the PkgServer would just be serving as a mirror and the standard format for mirrors is usually just arbitrary files. That's how I assumed it would work when I first saw the `Artifacts.toml` format, anyway. It looks like a format for specifying "here's how to download an arbitrary sequence of bytes with a given `sha256` hash."

**Comment by Keno** [comment]
We want to be able to do things like serve incremental updates as binary diffs and bundle commonly requested artifacts together.

**Comment by fredrikekre** [comment]
https://github.com/JuliaLang/Pkg.jl/issues/1921

**Comment by thomasjm** [comment]
Oh, thanks @fredrikekre ! 

I had just typed a follow-up to @Keno -- would you mind answering one more round of questions here so I understand the context better? Then I'll shift over to JuliaLang/Pkg.jl#1921.

### Original response: 

Oh, like you would try to detect if a single file in a tar archive changes and then only send a diff of that file?

FWIW I'm coming from a [Nix](https://nixos.org/) packaging perspective so that might color the way I see this--

I don't follow how bundling commonly requested artifacts together requires them to be tar archives, unless you mean to take N tar archive artifacts and compress them all into a single big tar archive artifact. This seems unlikely to save much space on compression compared to just doing N separate downloads. I might not be getting this so I'll mainly respond to the incremental update part...

It seems to me that you don't need to bake in the assumption that the artifact is an archive file to do incremental updates. For one thing, binary blob diffing tools do exist and I believe they can even do efficient incremental updates for some archive types (i.e. zip). Or, you could detect which files appear to be archives and do incremental updates with those, while doing full updates on other binary blobs (for which incremental updates may not be efficient anyway).

TL;DR since the artifact system design is just about fully capable of supporting arbitrary files already, I would suggest that trying to make them possible would have big benefits for package authors, and probably wouldn't prevent the optimizations you want to do. But still trying to understand fully :)

**Comment by Keno** [comment]
It's not a problem of technical impossibility, it's a problem of complexity. Right now the semantics of artifacts are very simple "a content addressed file tree". Expanding the semantics to say " a content addressed file tree or a single file" adds significant complexity. Suddenly you need out-of-band storage for the file name, file attributes, etc, as well as add an extra case to everything inside the infrastructure that needs to deal with artifacts. Maybe that would be worth it for a compelling use case, but the use case presented fails to meet that bar. We've had bad experiences with downloading dependencies from arbitrary upstream websites, because they tend to have a bit of a loose interpretation of backwards compatibility and reliability. Maybe that doesn't apply to this particular upstream, but it is nevertheless a concern in general, so further encouraging such reliance is not compelling. Perhaps the solution here is simply a nice webform that you give an existing URL and it dumps it into a minimal tarball for you and uploads it to GitHub releases or whatever.

**Comment by thomasjm** [comment]
Got it, thanks for explaining!
